### PR TITLE
Synchronize attribute properties

### PR DESCRIPTION
**Pull request by kswenson**
Synchronize attribute properties (formula, description, unit, precision, etc.)

cf. [PT #165131819](https://www.pivotaltracker.com/story/show/165131819)

### CODE REVIEW

File: src/lib/codap-helper.ts [hunk 1]
```diff
@@ -314,17 +332,45 @@ export class CodapHelper {
...
+      // synchronize properties of existing attributes
+      originalAttributes.forEach(origAttr => {
+        const sharedAttr = sharedAttributes.find(attr => attr.name === origAttr.name);
+        if (sharedAttr) {
+          const origAttrProps = origAttr.attr as any;
+          const sharedAttrProps = sharedAttr.attr as any;
+          const propsToUpdate: any = {};
+          let changed = false;
+          // tslint:disable-next-line: forin
+          for (const prop in sharedAttrProps) {
```
Comment (sfentress) [inline src/lib/codap-helper.ts]: Is there any chance that the user's original attributes have props that are unset/non-existent in the incoming attribute, and we would want to change the original attribute to match?

Would it be simpler and more robust to do a `JSON.stringify` comparison, and, if there are any differences, simple set the whole attribute to the new props, and not worry about which specific props have changed?

Comment (kswenson) [reply src/lib/codap-helper.ts]: Good question -- I thought about that approach and it felt a bit heavy-handed, but maybe that's appropriate in this case.

**Comment by kswenson** [comment]
After some further testing, the order of the properties is not deterministic, so the JSON.stringified contents rarely (if ever) match for uninteresting reasons. I did remove the null test, however, since as you suggested there are theoretically cases in which a null/undefined value should replace a non-null/undefined value. That said, I don't think firebase can return a null/undefined value in this scenario, so it may not matter in this case.

**Comment by sfentress** [comment]
That's surprising to me. JSON.stringify is supposed to be at minimum internally-consistent, so two identical objects are always expected to return the same string for the same browser.

As for the null props, what I was referring to was actually props that are simply missing in `sharedAttrProps`. E.g., if (contrived example), we had our original attribute:

```
{
  id: "attr1",
  isLocked: true
}
```
and our incoming shared attribute
```
{
  id: "attr1"
}
```
the code above wouldn't remove `isLocked` from our original attribute. I don't know if CODAP has any values that aren't serialized when they're not the default.

**Comment by sfentress** [comment]
Actually, the `Formula` property does indeed work like that. As the code stands now, adding a formula to an attribute will add a new prop to the attribute object, so the sharing will work as expected, but deleting the formula will simply remove the prop, and so the code `for (const prop in sharedAttrProps)` will simply miss it, and the formula will not be removed for the other users.

**Comment by kswenson** [comment]
[MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify):
>Note: Properties of non-array objects are not guaranteed to be stringified in any particular order. Do not rely on ordering of properties within the same object within the stringification.

I think it would be the same if the same object were stringified twice, but not for two different objects with the same properties, particularly if the properties were added in a different order.

Your other point is well taken. CODAP is loose with distinctions between null/empty string/undefined, while firebase treats them differently, so I'll need to handle that case. Thanks

**Pull request merged**